immuni_common/helpers/utils.py

```diff
@@ -52,3 +53,32 @@ def dense_dict(dictionary: Dict) -> Dict:
         for k, v in dictionary.items()
         if v is not None
     }
+
+
+# Note: This could be a Generic[T] with T as TypeVar. That would allow
+#  us to define a WeightedPair[str] and have the weighted_random
+#  return a valid typed str when used, but generic NamedTuples are not yet
+#  supported by mypy:
+#  https://github.com/python/mypy/issues/685
+class WeightedPair(NamedTuple):
+    """
+    Simple NamedTuple that allows us to describe
+    weighted key/value pairs for weighted random.
+    """
+
+    weight: int
+    payload: Any
+
+
+def weighted_random(pairs: List[WeightedPair]) -> Any:
+    """
+    Returns the value associated with
+    """
+    pairs = [p for p in pairs if p.weight > 0]
```

Comment:
I'd raise if any of the values is `<=0`. I don't see the point in passing such values.

Reply:
I'd raise also if the set of payloads is not unique, something like:
```python
payloads = [p.payload for p in pairs]
unique_payloads = set(payloads)
if len(unique_payloads) != len(payloads):
    raise ImmuniException(f"Duplicate payloads in {payloads})
```

Reply:
But then it would become untestable 😞

Reply:
I didn't get the untestable part, sorry. 😞

Reply:
If we raise whenever we use a `0` weight, we cannot create a reliable test case (because we can't use 0-probabilities).

In that case the best we could do would be to run the test N times and make sure that we received the correct responses proportionally to the weights that we provided, but even then, we'd still be introducing a non-null probability that the test fails (for random chance).

As of now, we use 0-probabilities to make sure tests are predictable and cannot fail.

Reply:
Let me sum it up:
* we should allow the `weight == 0` for reproducible and easy testing (I'd add a note)
* we should raise if `weight < 0`
* we should raise if `payloads are not unique`.

Does this make sense?

Reply:
Why raise if payloads are not unique? That would be up to the caller's responsibility.

I agree on the rest.